tensorboard/examples/plugins/example_raw_scalars/tensorboard_plugin_example_raw_scalars/plugin.py

```diff
@@ -84,12 +86,23 @@ def _serve_static_file(self, request):
         Requests from the frontend have a path in this form:
         /data/plugin/example_raw_scalars/static/foo
         This serves the appropriate asset: ./static/foo.
+
+        Checks the normpath to guard against path traversal attacks.
         """
         static_path_part = request.path[len(_PLUGIN_DIRECTORY_PATH_PART) :]
-        res_path = os.path.join(os.path.dirname(__file__), static_path_part)
 
-        with open(res_path, "rb") as read_file:
-            mimetype = mimetypes.guess_type(res_path)[0]
+        posix_path = static_path_part.replace(ntpath.sep, posixpath.sep)
+        required_prefix = posixpath.join(posixpath.dirname(__file__), "static/")
+        full_path = posixpath.join(posixpath.dirname(__file__), posix_path)
+        norm_path = posixpath.normpath(full_path)
+
+        if not norm_path.startswith(required_prefix):
```

Comment:
Okay, I think that this is safe, but I don’t see how it reads the
correct file. On Windows, `__file__` is like `C:\Users\someone\...`, but
the `posixpath.dirname` of that value is the empty string, because it
contains no POSIX path separators. So `norm_path` ends up being just the
NT-to-POSIX-converted input path without any directory prefix at all,
which behaves correctly under the safety check but won’t actually be
read from the filesystem correctly (because it will be a relative path,
and the current working directory will not be `dirname(__file__)`).

Do you think that it would make more sense to use `os.path.normpath`,
given that we’re trying to sanitize access to the platform-specific file
system? Something like:

```python
resource_name = os.path.normpath(static_path_part)
if not resource_name.startswith("static" + os.path.sep):
  return ...
resource_path = os.path.join(os.path.dirname(__file__), resource_name)
with open(resource_path, "rb") as infile:
  contents = infile.read()
...
```

I had hoped that `posixpath.normpath` could make this behavior more
portable, but I guess given a URL with encoded backslashes there isn’t
really a clearly portable way to handle it (that I can see—?).

Reply:
You're correct, it didn't work on Windows, but it was safe.

I haven't come up with a better solution, so I've gone with your suggestion.  It looks safe, and the only strange thing is that behavior may differ between platforms if the user want to serve a file with a name that includes backslashes.  I've added a test case for this, and that behavior seems acceptable to me.

Once again, I'll update this thread when I've tested that it works on a Windows machine.

Reply:
Can confirm this plugin works on Windows 10, Python 3.6.8